controllers/bmcjob_controller.go

```diff
@@ -18,38 +18,207 @@ package controllers
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+
+		logger.Error(err, "Failed to get BMCJob")
+		return ctrl.Result{}, err
+	}
+
+	// Deletion is a noop.
+	if !bmcJob.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+
+	// Job has StartTime is noop.
+	if !bmcJob.Status.StartTime.IsZero() {
+		return ctrl.Result{}, nil
+	}
+	// Create a patch from the initial BMCJob object
+	// Patch is used to update Status after reconciliation
+	bmcJobPatch := client.MergeFrom(bmcJob.DeepCopy())
+
+	return r.reconcile(ctx, bmcJob, bmcJobPatch, logger)
+}
+
+func (r *BMCJobReconciler) reconcile(ctx context.Context, bmj *bmcv1alpha1.BMCJob, bmjPatch client.Patch, logger logr.Logger) (ctrl.Result, error) {
+	// Get BaseboardManagement object for the Job
+	// Requeue if error
+	baseboardManagement := &bmcv1alpha1.BaseboardManagement{}
+	err := r.resolveBaseboardManagementRef(ctx, bmj, baseboardManagement)
+	if err != nil {
+		return ctrl.Result{Requeue: true}, fmt.Errorf("resolving BMCJob %s/%s BaseboardManagementRef: %v", bmj.Namespace, bmj.Name, err)
+	}
+
+	// Fetching username, password from SecretReference
+	// Requeue if error fetching secret
+	username, password, err := resolveAuthSecretRef(ctx, r.client, baseboardManagement.Spec.Connection.AuthSecretRef)
+	if err != nil {
+		return ctrl.Result{Requeue: true}, fmt.Errorf("resolving BaseboardManagement %s/%s SecretReference: %v", baseboardManagement.Namespace, baseboardManagement.Name, err)
+	}
+
+	// Initializing BMC Client
+	bmcClient, err := r.bmcClientFactory(ctx, baseboardManagement.Spec.Connection.Host, strconv.Itoa(baseboardManagement.Spec.Connection.Port), username, password)
+	if err != nil {
+		logger.Error(err, "BMC connection failed", "host", baseboardManagement.Spec.Connection.Host)
+		bmj.SetCondition(bmcv1alpha1.JobFailed, bmcv1alpha1.BMCJobConditionTrue, bmcv1alpha1.WithJobConditionMessage(fmt.Sprintf("Failed to connect to BMC: %v", err)))
+		result, patchErr := r.patchBMCJobStatus(ctx, bmj, bmjPatch)
+		if patchErr != nil {
+			return result, utilerrors.NewAggregate([]error{patchErr, err})
+		}
+
+		return result, err
+	}
+
+	// Pause BaseboardManagement reconciliation before running BMCJob tasks
+	baseboardManagement.PauseReconcile()
+
+	defer func() {
+		// Close BMC connection after reconcilation
+		err = bmcClient.Close(ctx)
+		if err != nil {
+			logger.Error(err, "BMC close connection failed", "host", baseboardManagement.Spec.Connection.Host)
+		}
+		// Delete the BaseboardManagement pause annotation
+		baseboardManagement.ClearPauseAnnotation()
+	}()
+
+	return r.reconcileBMCTasks(ctx, bmj, bmjPatch, bmcClient, logger)
+}
+
+// reconcileBMCTasks Creates the individual BMCTasks
+// Runs each of the BMCTasks
+func (r *BMCJobReconciler) reconcileBMCTasks(ctx context.Context, bmj *bmcv1alpha1.BMCJob, bmjPatch client.Patch, bmcClient BMCClient, logger logr.Logger) (ctrl.Result, error) {
```

Comment:
The BMCJob reconciliation just needs to take the `Tasks` field and create `BMCTask`s to be run by the `BMCTask` controller. Alternatively we simplify everything and have the job controller run the tasks but we had good reasons, which escape me right now, for having a task CRD. Lets have a chat offline.